### PR TITLE
Colony flag banners

### PR DESCRIPTION
**Pull request by ShadowProtocol**
# Changes proposed in this pull request:
- Added a custom banner picker window to select pattern designs, accessible from the town hall gui
- Added the colony flag banner custom banner blocks (one normal, one for walls) 
    - Can right click on a banner with a colony flag banner to copy its design as a default
    - Colony flag banner displays the colony flag set in the town hall, or the default if none was set (or there is no colony)
    - Each banner renders a small solid substitution block in the center while in creative mode, to tell them apart from normal banners.
    - Guards now apply the banner patterns for the colony flag to their shields (only updated when blocking, though)

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/blocks/decorative/BlockColonyFlagWallBanner.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Override
+    public VoxelShape getShape(BlockState state, IBlockReader worldIn, BlockPos pos, ISelectionContext context)
+    {
+        return BANNER_SHAPES.get(state.get(HORIZONTAL_FACING));
+    }
+
+    @Override
+    public BlockState getStateForPlacement(BlockItemUseContext context) {
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/blocks/decorative/BlockColonyFlagWallBanner.java]: formatting

File: src/main/java/com/minecolonies/coremod/blocks/decorative/BlockColonyFlagBanner.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.util.math.shapes.ISelectionContext;
+import net.minecraft.util.math.shapes.VoxelShape;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
+import net.minecraft.world.IWorldReader;
+
+import java.util.Map;
+
+public class BlockColonyFlagBanner extends AbstractColonyFlagBanner<BlockColonyFlagBanner>
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/blocks/decorative/BlockColonyFlagBanner.java:30]: jdoc maybe

File: src/main/java/com/minecolonies/coremod/blocks/decorative/BlockColonyFlagWallBanner.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.shapes.ISelectionContext;
+import net.minecraft.util.math.shapes.VoxelShape;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
+import net.minecraft.world.IWorldReader;
+
+import java.util.Map;
+
+public class BlockColonyFlagWallBanner extends AbstractColonyFlagBanner<BlockColonyFlagWallBanner>
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/blocks/decorative/BlockColonyFlagWallBanner.java:27]: jdoc

File: src/main/java/com/minecolonies/coremod/client/render/TileEntityColonyFlagRenderer.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import net.minecraft.item.DyeColor;
+import net.minecraft.item.ItemStack;
+import net.minecraft.tileentity.BannerPattern;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.world.GameType;
+
+import java.util.List;
+
+public class TileEntityColonyFlagRenderer extends TileEntityRenderer<TileEntityColonyFlag>
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/client/render/TileEntityColonyFlagRenderer.java:35]: some jdoc

File: src/main/java/com/minecolonies/coremod/items/ItemColonyFlagBanner.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import net.minecraft.tileentity.BannerTileEntity;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.ActionResultType;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+public class ItemColonyFlagBanner extends BannerItem
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/items/ItemColonyFlagBanner.java:30]: jdoc

File: src/main/java/com/minecolonies/coremod/network/messages/server/colony/ColonyFlagChangeMessage.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import com.minecolonies.api.util.constant.Constants;
+import com.minecolonies.coremod.Network;
+import com.minecolonies.coremod.network.messages.server.AbstractColonyServerMessage;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.nbt.ListNBT;
+import net.minecraft.network.PacketBuffer;
+import net.minecraftforge.fml.network.NetworkEvent;
+import org.apache.logging.log4j.LogManager;
+
+public class ColonyFlagChangeMessage extends AbstractColonyServerMessage
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/network/messages/server/colony/ColonyFlagChangeMessage.java:18]: jdoc all over this =D

File: src/api/java/com/minecolonies/api/colony/IColony.java [hunk 7]
```diff
@@ -300,6 +311,8 @@ default String getTeamName()
 
     void setColonyColor(TextFormatting color);
 
+    void setColonyFlag(ListNBT patterns);
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/colony/IColony.java:314]: jdoc

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBannerPicker.java [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+
+        RenderHelper.setupGuiFlatDiffuseLighting();
+        drawFlag();
+
+        super.render(mouseX, mouseY, partialTicks);
+
+        // Render the instructions
+        this.drawCenteredString(
+                this.font,
+                "Choose a colored pattern for up to 6 layers on your flag.",
```
Comment (MatthiasMann) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBannerPicker.java]: Take text from the language file?

Comment (ShadowProtocol) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowBannerPicker.java]: You're looking at an older commit.

Comment (MatthiasMann) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowBannerPicker.java]: Ohh sorry

File: src/api/java/com/minecolonies/api/colony/IColony.java (hunk 7 above)
Comment (ShadowProtocol) [reply src/api/java/com/minecolonies/api/colony/IColony.java:314]: This is a setter, and few of the others have docs. I wouldn't want it to be alone ...

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/5506/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

File: src/main/java/com/minecolonies/coremod/colony/ColonyView.java [hunk 9]
```diff
@@ -289,6 +304,10 @@ public static void serializeNetworkData(@NotNull Colony colony, @NotNull PacketB
 
         buf.writeInt(colony.getTeamColonyColor().ordinal());
 
+        CompoundNBT flagNBT = new CompoundNBT();
+        flagNBT.put("Patterns", colony.getColonyFlag());
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/ColonyView.java]: constant for read and write string key

**Review by someaddons — COMMENTED** [review]
Looks nice, just should add a constant for the "Patterns" nbt key you used at 5-6 places

**Pull request merged**